eppo_core/src/event_ingestion/event_ingestion.rs

```diff
@@ -0,0 +1,228 @@
+use std::time::Duration;
+
+use tokio::sync::mpsc;
+use url::Url;
+use uuid::Uuid;
+
+use crate::{background::BackgroundRuntime, sdk_key::SdkKey};
+
+use super::{
+    auto_flusher, batcher, delivery, event_delivery::EventDelivery, queued_event::QueuedEvent,
+    retry, BatchedMessage, Event,
+};
+
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct EventIngestionConfig {
+    pub sdk_key: SdkKey,
+    pub ingestion_url: Url,
+    pub delivery_interval: Duration,
+    pub retry_interval: Duration,
+    pub max_retry_delay: Duration,
+    pub max_retries: u32,
+    pub batch_size: usize,
+    pub max_queue_size: usize,
+}
+
+impl EventIngestionConfig {
+    const DEFAULT_BATCH_SIZE: usize = 1_000;
+    const DEFAULT_DELIVERY_INTERVAL: Duration = Duration::from_secs(10);
+    const DEFAULT_RETRY_INTERVAL: Duration = Duration::from_secs(5);
+    const DEFAULT_MAX_RETRY_DELAY: Duration = Duration::from_secs(30);
+    const DEFAULT_MAX_RETRIES: u32 = 3;
+    const DEFAULT_MAX_QUEUE_SIZE: usize = 10_000;
+
+    /// Creates new event ingestion config.
+    ///
+    /// Returns `None` if `sdk_key` is not suitable for event ingestion.
+    pub fn new(sdk_key: SdkKey) -> Option<Self> {
+        let ingestion_url = sdk_key.event_ingestion_url()?;
+        let config = EventIngestionConfig {
+            sdk_key,
+            ingestion_url,
+            delivery_interval: Self::DEFAULT_DELIVERY_INTERVAL,
+            retry_interval: Self::DEFAULT_RETRY_INTERVAL,
+            max_retry_delay: Self::DEFAULT_MAX_RETRY_DELAY,
+            max_retries: Self::DEFAULT_MAX_RETRIES,
+            batch_size: Self::DEFAULT_BATCH_SIZE,
+            max_queue_size: Self::DEFAULT_MAX_QUEUE_SIZE,
+        };
+        Some(config)
+    }
+
+    pub fn spawn(&self, runtime: &BackgroundRuntime) -> EventIngestion {
+        EventIngestion::spawn(runtime, self)
+    }
+}
+
+/// A handle to Event Ingestion subsystem.
+pub struct EventIngestion {
+    tx: mpsc::Sender<BatchedMessage<QueuedEvent>>,
+}
+
+impl EventIngestion {
+    /// Starts the event ingestion subsystem on the given background runtime.
+    pub fn spawn(runtime: &BackgroundRuntime, config: &EventIngestionConfig) -> EventIngestion {
+        let event_delivery =
+            EventDelivery::new(config.sdk_key.clone(), config.ingestion_url.clone());
+
+        let (input, flusher_uplink) = mpsc::channel(config.max_queue_size);
+        let (flusher_downlink, batcher_uplink) = mpsc::channel(1);
+        let (batcher_downlink, delivery_uplink) = mpsc::channel(1);
+        let (delivery_downlink, retry_uplink) = mpsc::channel(1);
+        let (delivery_status_tx, delivery_status_rx) = mpsc::channel(1);
```

Comment:
@rasendubi in a different PR you mentioned you wanted to bump one of these to avoid deadlocks, still planning it for the future?

Reply:
Yes! It needs to be unbound, which is a different channel type, so needs changes to both retry and delivery components.

I also have a thought for another solution. Given how simple retry technically is, we could just spawn one-off delay-then-send tasks on tokio runtime directly from delivery, and that would remove the loop (and thus deadlock) completely